File: ios/App/App/NativeAudioPlugin.swift
import Foundation
import Capacitor
import AVFoundation
import MediaPlayer

@objc(NativeAudioPlugin)
public class NativeAudioPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeAudioPlugin"
    public let jsName = "NativeAudio"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "prepare", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "play", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pause", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "seek", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getProgress", returnType: CAPPluginReturnPromise)
    ]

    private var queuePlayer: AVQueuePlayer?
    private var introItem: AVPlayerItem?
    private var mainItem: AVPlayerItem?
    
    // Crossfade observer token
    private var timeObserverToken: Any?
    private var isFading = false

    override public func load() {
        do {
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print("Failed to set audio session category: \(error)")
        }
        setupRemoteCommandCenter()
    }

    @objc func prepare(_ call: CAPPluginCall) {
        let introUrlStr = call.getString("introUrl")
        guard let mainUrlStr = call.getString("mainUrl"),
              let mainUrl = URL(string: mainUrlStr) else {
            call.reject("Invalid Main URL")
            return
        }

        let title = call.getString("title") ?? "Áudio"
        let author = call.getString("author") ?? "APP.PRIME"

        if let introUrlStr = introUrlStr, let introUrl = URL(string: introUrlStr) {
            introItem = AVPlayerItem(url: introUrl)
        } else {
            introItem = nil
        }
        
        mainItem = AVPlayerItem(url: mainUrl)

        if queuePlayer == nil {
            queuePlayer = AVQueuePlayer()
        } else {
            queuePlayer?.removeAllItems()
        }

        if let intro = introItem, let main = mainItem {
            queuePlayer?.insert(intro, after: nil)
            queuePlayer?.insert(main, after: intro)
        } else if let main = mainItem {
            queuePlayer?.insert(main, after: nil)
        }

        setupNowPlaying(title: title, author: author)

        call.resolve(["success": true])
    }

    @objc func play(_ call: CAPPluginCall) {
        queuePlayer?.play()
        updateNowPlayingPlaybackState()
        call.resolve()
    }

    @objc func pause(_ call: CAPPluginCall) {
        queuePlayer?.pause()
        updateNowPlayingPlaybackState()
        call.resolve()
    }

    @objc func seek(_ call: CAPPluginCall) {
        let time = call.getDouble("time") ?? 0
        
        guard let currentItem = queuePlayer?.currentItem else {
            call.resolve()
            return
        }

        // Simplesmente foca no item atual. (Para crossfade perfeito precisaria calcular se cai na intro ou no main)
        // Para simplificar, assumimos seek no item atual (que geralmente é o Main quando o usuário usa a barra)
        let cmTime = CMTime(seconds: time, preferredTimescale: 1000)
        currentItem.seek(to: cmTime, completionHandler: { _ in
            call.resolve()
        })
    }

    @objc func stop(_ call: CAPPluginCall) {
        queuePlayer?.pause()
        queuePlayer?.removeAllItems()
        MPNowPlayingInfoCenter.default().nowPlayingInfo = nil
        call.resolve()
    }

    @objc func getProgress(_ call: CAPPluginCall) {
        let currentTime = queuePlayer?.currentTime().seconds ?? 0
        let duration = queuePlayer?.currentItem?.duration.seconds ?? 0
        let isPlaying = queuePlayer?.rate != 0
        let currentItem = queuePlayer?.currentItem
        let trackIndex = (currentItem == introItem && introItem != nil) ? 0 : 1

        call.resolve([
            "currentTime": currentTime.isNaN ? 0 : currentTime,
            "duration": duration.isNaN ? 0 : duration,
            "isPlaying": isPlaying,
            "trackIndex": trackIndex
        ])
    }

    private func setupNowPlaying(title: String, author: String) {
        var nowPlayingInfo = [String: Any]()
        nowPlayingInfo[MPMediaItemPropertyTitle] = title
        nowPlayingInfo[MPMediaItemPropertyArtist] = author
        MPNowPlayingInfoCenter.default().nowPlayingInfo = nowPlayingInfo
    }

    private func updateNowPlayingPlaybackState() {
        var nowPlayingInfo = MPNowPlayingInfoCenter.default().nowPlayingInfo ?? [String: Any]()
        let isPlaying = queuePlayer?.rate != 0
        let currentTime = queuePlayer?.currentTime().seconds ?? 0
        let duration = queuePlayer?.currentItem?.duration.seconds ?? 0
        
        nowPlayingInfo[MPNowPlayingInfoPropertyElapsedPlaybackTime] = currentTime
        nowPlayingInfo[MPMediaItemPropertyPlaybackDuration] = duration
        nowPlayingInfo[MPNowPlayingInfoPropertyPlaybackRate] = isPlaying ? 1.0 : 0.0
        
        MPNowPlayingInfoCenter.default().nowPlayingInfo = nowPlayingInfo
    }

    private func setupRemoteCommandCenter() {
        let commandCenter = MPRemoteCommandCenter.shared()

        commandCenter.playCommand.addTarget { [weak self] event in
            self?.queuePlayer?.play()
            self?.updateNowPlayingPlaybackState()
            return .success
        }

        commandCenter.pauseCommand.addTarget { [weak self] event in
            self?.queuePlayer?.pause()
            self?.updateNowPlayingPlaybackState()
            return .success
        }
    }
}

File: android/app/src/main/java/br/com/app/gpu2675756/gpu0e7509bfb7bde52aef412888bb17a456/NativeAudioPlugin.kt
package br.com.app.gpu2675756.gpu0e7509bfb7bde52aef412888bb17a456

import android.media.AudioAttributes
import android.media.MediaPlayer
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeAudio")
class NativeAudioPlugin : Plugin() {

    private var mediaPlayerIntro: MediaPlayer? = null
    private var mediaPlayerMain: MediaPlayer? = null
    private var isIntroPlaying = true

    @PluginMethod
    fun prepare(call: PluginCall) {
        val introUrl = call.getString("introUrl")
        val mainUrl = call.getString("mainUrl")

        if (introUrl == null || mainUrl == null) {
            call.reject("Invalid URLs")
            return
        }

        try {
            mediaPlayerIntro?.release()
            mediaPlayerMain?.release()

            mediaPlayerIntro = MediaPlayer().apply {
                setAudioAttributes(
                    AudioAttributes.Builder()
                        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .build()
                )
                setDataSource(introUrl)
                prepareAsync()
            }

            mediaPlayerMain = MediaPlayer().apply {
                setAudioAttributes(
                    AudioAttributes.Builder()
                        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .build()
                )
                setDataSource(mainUrl)
                prepareAsync()
            }

            // Quando a intro terminar, toca a principal
            mediaPlayerIntro?.setOnCompletionListener {
                isIntroPlaying = false
                mediaPlayerMain?.start()
            }

            call.resolve(JSObject().put("success", true))
        } catch (e: Exception) {
            call.reject("Error setting up MediaPlayer: ${e.message}")
        }
    }

    @PluginMethod
    fun play(call: PluginCall) {
        if (isIntroPlaying) {
            mediaPlayerIntro?.start()
        } else {
            mediaPlayerMain?.start()
        }
        call.resolve()
    }

    @PluginMethod
    fun pause(call: PluginCall) {
        if (isIntroPlaying) {
            mediaPlayerIntro?.pause()
        } else {
            mediaPlayerMain?.pause()
        }
        call.resolve()
    }

    @PluginMethod
    fun seek(call: PluginCall) {
        val time = call.getDouble("time") ?: 0.0
        if (!isIntroPlaying) {
            mediaPlayerMain?.seekTo((time * 1000).toInt())
        }
        call.resolve()
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        mediaPlayerIntro?.stop()
        mediaPlayerMain?.stop()
        call.resolve()
    }

    @PluginMethod
    fun getProgress(call: PluginCall) {
        val ret = JSObject()
        if (isIntroPlaying) {
            ret.put("currentTime", (mediaPlayerIntro?.currentPosition ?: 0) / 1000.0)
            ret.put("duration", (mediaPlayerIntro?.duration ?: 0) / 1000.0)
            ret.put("isPlaying", mediaPlayerIntro?.isPlaying ?: false)
            ret.put("trackIndex", 0)
        } else {
            ret.put("currentTime", (mediaPlayerMain?.currentPosition ?: 0) / 1000.0)
            ret.put("duration", (mediaPlayerMain?.duration ?: 0) / 1000.0)
            ret.put("isPlaying", mediaPlayerMain?.isPlaying ?: false)
            ret.put("trackIndex", 1)
        }
        call.resolve(ret)
    }
}
